pkg/monitoring/mimir/service.go

```diff
@@ -0,0 +1,163 @@
+package mimir
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/giantswarm/observability-operator/pkg/common"
+	"github.com/giantswarm/observability-operator/pkg/common/password"
+	"github.com/giantswarm/observability-operator/pkg/common/secret"
+	"github.com/giantswarm/observability-operator/pkg/monitoring/prometheusagent"
+)
+
+const (
+	ingressAuthSecretName = "mimir-gateway-ingress-auth"
+	mimirApiKey           = "mimir-basic-auth"
+	mimirNamespace        = "mimir"
+)
+
+type MimirService struct {
+	client.Client
+	PasswordManager password.Manager
+	common.ManagementCluster
+}
+
+// ConfigureMimir configures the ingress and its authentication (basic auth)
+// to allow prometheus agents to send their data to Mimir
+func (ms *MimirService) ConfigureMimir(ctx context.Context) error {
+	logger := log.FromContext(ctx).WithValues("cluster", ms.ManagementCluster.Name)
+	logger.Info("configuring mimir ingress")
+
+	err := ms.CreateApiKey(ctx, logger)
+	if err != nil {
+		logger.Error(err, "failed to create mimir auth secret")
+		return errors.WithStack(err)
+	}
+
+	err = ms.CreateIngressAuthenticationSecret(ctx, logger)
+	if err != nil {
+		logger.Error(err, "failed to create mimir ingress secret")
+		return errors.WithStack(err)
+	}
+
+	logger.Info("configured mimir ingress")
+
+	return nil
+}
+
+func (ms *MimirService) CreateApiKey(ctx context.Context, logger logr.Logger) error {
+	objectKey := client.ObjectKey{
+		Name:      mimirApiKey,
+		Namespace: mimirNamespace,
+	}
+
+	current := &corev1.Secret{}
+	err := ms.Client.Get(ctx, objectKey, current)
+	if apierrors.IsNotFound(err) {
+		// First all secrets using the password from the mimirApiKey secret are deleted
+		// to ensure that they won't use an outdated password.
+		logger.Info("Deleting old secrets")
+
+		err := secret.DeleteSecret(ingressAuthSecretName, mimirNamespace, ctx, ms.Client)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		clusterList := &clusterv1.ClusterList{}
+		err = ms.Client.List(ctx, clusterList)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		for _, cluster := range clusterList.Items {
+			secretName := prometheusagent.GetPrometheusAgentRemoteWriteSecretName(&cluster)
+			err = secret.DeleteSecret(secretName, cluster.Namespace, ctx, ms.Client)
```

Comment:
How do we make sure those secrets are not re-created afterwards ?

Reply:
Check what I answered Quentin ;) 

> Well, since this deletion is happening in the case where the basic-auth secret isn't present on the cluster and considering that the remotewrite secrets and ingress secret all need the basic-auth secret's data and can't be created without it, as long as this process is not finished and the basic-auth created, they won't be created before.
Not sure if I explained it clearly 😅

Reply:
I did not understand that ^

Reply:
i'll try to rephrase : 

basically, remotewrite secrets as well as the secret used by the ingress rely on the mimirApiKey secret's data (whose value is `username:password`) for their creation process meaning that as long as the mimirApiKey secret is not created, neither can they. 
Since the piece of code you mentioned is happening **prior** to the mimirApiKey secret being created, one can be sure that all other secrets won't be created before this piece of code has been executed.

Reply:
If I got this right, here we are deleting the remotewrite secret, and mimirApiKey creation happens later on right?
What happen then when we delete the secret here and the mimiApiKey creation is triggered ? Isn't this causing a failure ?